### ISSUE TITLE
core dump in /...(?{ code })

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4673]
Migrated from [rt.perl.org#8027](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=8027) (status was 'resolved')

Searchable as RT8027$

**Comment by p5pRT** [comment]
### From john@pc.xs4all.nl

#### Created by john@pc.xs4all.nl


This is a bug report for perl from john@&#8203;pc\.xs4all\.nl\,
generated with the help of perlbug 1\.33 running under perl v5\.7\.2\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
In some circumstances\, perl dumps core after executing code embedded
in a regular expression\, with \(?\{ code\.\.\. \}\)\. The following program is
an example of this\.

The crash happens with perl 5\.6\.1\, perl 5\.7\.2\, and the latest bleadperl\.
The code doesn't compile with perl 5\.005\_03\, the \(?\{ code \}\) construct
wasn't mature enough back then\, apparently\.

I haven't really tried to minimize this\. If this is a known issue\, I'd
rather not waste too much time on it\. However if someone is interested
in fixing this\, I can try to reduce the example to a bare minimum\.

The workaround is to move code out of the regex itself\, and only do
a minimum amount of work inside the re\. So I did manage to match the
$sillytimestamp\, using a slightly modified version\.

Here is the program&#8203;:

\-\-\-\-8\<\-BEGIN\-PROGRAM\-\-\-\-\-\-\-
\#\!/usr/bin/perl

use Time&#8203;::Local;
use strict;

use re 'debug';

sub match \{
&nbsp;    my $str = shift;
&nbsp;    $str =~ m\{
&nbsp;        ^
&nbsp;        \(\\d|\[1\-5\]\\d\)        \# seconds
&nbsp;        \(\\d|\[1\-5\]\\d\)        \# minutes
&nbsp;        \(\\d|1\\d|2\[0\-3\]\)        \# hours
&nbsp;        \(\[1\-9\]|\[12\]\\d|3\[01\]\)    \# mday
&nbsp;        \(\\d|1\[01\]\)            \# mon
&nbsp;        \(\[2\-9\]|\[1\-9\]\\d|1\[0\-2\]\\d|13\[0\-7\]\)    \# year
&nbsp;        \(\[0\-6\]\)            \# wday
&nbsp;        \(\\d|\[1\-9\]\\d|\[12\]\\d\\d|3\[0\-5\]\\d|36\[0\-5\]\)    \# yday
&nbsp;        \(\[01\]\)            \# isdst
&nbsp;        $
&nbsp;        \(?\{
&nbsp;        print "Possible match&#8203;: $1\-$2\-$3\-$4\-$5\-$6\-$7\-$8\-$9\\n";
&nbsp;        my $t = timelocal\($1\, $2\, $3\, $4\, $5\, 1900 \+ $6\);
&nbsp;        print "Possible date&#8203;: $t\\n";
&nbsp;        my @&#8203;lt = localtime\($t\);
&nbsp;        print "Localtime is @&#8203;lt\\n";
&nbsp;        if \( $lt\[6\] == $7 && $lt\[7\] == $8 && $lt\[8\] == $9 \) \{
&nbsp;            print "MATCH\! "\, scalar localtime $t\, "\\n";
&nbsp;        \}
&nbsp;        \}\)
&nbsp;     \}x;
\}

my $sillytimestamp = "251013211010133240";
match\($sillytimestamp\);
\-\-\-\-END\-PROGRAM\->8\-\-\-\-\-\-\-\-

This is \(a portion of\) the output\, when run under yesterday's bleadperl
compiled with \-DDEBUGGING\.

\-\-\-\-8\<\-BEGIN\-OUTPUT\-\-\-\-\-\-\-
pc&#8203;:\~/perl/bleadperl $ \./perl \-Ilib \-w \.\./localtime\_play/recrash 
Compiling REx \`
&nbsp;            ^
&nbsp;            \(\\d|\[1\-5\]\\d\)                \# seconds
&nbsp;            \(\\d|\[1\-5\]\\d\)                \# minutes
&nbsp;            \(\\d|1\\d|2\[0\-3\]\)             \# hours
&nbsp;            \(\[1\-9\]|\[12\]\\d|3\[01\]\)        \# mday
&nbsp;            \(\\d|1\[01\]\)                  \# mon
&nbsp;            \(\[2\-9\]|\[1\-9\]\\d|1\[0\-2\]\\d|13\[0\-7\]\)    \# year
&nbsp;            \(\[0\-6\]\)                     \# wday
&nbsp;            \(\\d|\[1\-9\]\\d|\[12\]\\d\\d|3\[0\-5\]\\d|36\[0\-5\]\)      \# yday
&nbsp;            \(\[01\]\)                      \# isdst
&nbsp;            $
&nbsp;            \(?\{
&nbsp;                print "Possible match&#8203;: $1\-$2\-$3\-$4\-$5\-$6\-$7\-$8\-$9\\n";
&nbsp;                my $t = timelocal\($1\, $2\, $3\, $4\, $5\, 1900 \+ $6\);
&nbsp;                print "Possible date&#8203;: $t\\n";
&nbsp;                my @&#8203;lt = localtime\($t\);
&nbsp;                print "Localtime is @&#8203;lt\\n";
&nbsp;                if \( $lt\[6\] == $7 && $lt\[7\] == $8 && $lt\[8\] == $9 \) \{
&nbsp;                    print "MATCH\! "\, scalar localtime $t\, "\\n";
&nbsp;                \}
&nbsp;            \}\)
&nbsp;         '
size 280 Got 2244 bytes for offset annotations\.
first at 2
rarest char 
 at 0
&nbsp;   1&#8203;: BOL\(2\)
\[\.\.\.\]
EXECUTING\.\.\.
\[\.\.\.\]
&nbsp;  17 \<21101013324> \<0>    |264&#8203;:                ANYOF\[01\]
&nbsp;  18 \<211010133240> \<>    |275&#8203;:                CLOSE9
&nbsp;  18 \<211010133240> \<>    |277&#8203;:                EOL
&nbsp;  18 \<211010133240> \<>    |278&#8203;:                EVAL
&nbsp;  re\_eval 0x81a6040
Possible match&#8203;: 25\-10\-13\-21\-10\-101\-3\-324\-0
Possible date&#8203;: 1006344625
Localtime is 25 10 13 21 10 101 3 324 0
MATCH\! Wed Nov 21 13&#8203;:10&#8203;:25 2001
&nbsp;  18 \<211010133240> \<>    |280&#8203;:                END
Match successful\!
Segmentation fault \(core dumped\)
\-\-\-\-END\-OUTPUT\->8\-\-\-\-\-\-\-\-

And here is what gdb says about the generated core file&#8203;:

pc&#8203;:\~/perl/bleadperl $ gdb perl core
GNU gdb 19991004
Copyright 1998 Free Software Foundation\, Inc\.
GDB is free software\, covered by the GNU General Public License\, and you are
welcome to change it and/or distribute copies of it under certain conditions\.
Type "show copying" to see the conditions\.
There is absolutely no warranty for GDB\.  Type "show warranty" for details\.
This GDB was configured as "i386\-redhat\-linux"\.\.\.
Core was generated by \`\./perl \-Ilib \-w \.\./localtime\_play/recrash'\.
Program terminated with signal 11\, Segmentation fault\.
Reading symbols from /lib/libnsl\.so\.1\.\.\.done\.
Reading symbols from /lib/libdl\.so\.2\.\.\.done\.
Reading symbols from /lib/libm\.so\.6\.\.\.done\.
Reading symbols from /lib/libc\.so\.6\.\.\.done\.
Reading symbols from /lib/libcrypt\.so\.1\.\.\.done\.
Reading symbols from /lib/libutil\.so\.1\.\.\.done\.
Reading symbols from /lib/ld\-linux\.so\.2\.\.\.done\.
Reading symbols from /usr/home/john/perl/bleadperl/lib/auto/re/re\.so\.\.\.done\.
\#0  0x80f3b9c in Perl\_leave\_scope \(base=13\) at scope\.c&#8203;:825
825                 if \(SvREFCNT\(sv\) \<= 1 && \!SvOBJECT\(sv\)\) \{
\(gdb\) bt
\#0  0x80f3b9c in Perl\_leave\_scope \(base=13\) at scope\.c&#8203;:825
\#1  0x80c5eb7 in Perl\_pp\_match \(\) at pp\_hot\.c&#8203;:1375
\#2  0x80ac7f4 in Perl\_runops\_debug \(\) at dump\.c&#8203;:1388
\#3  0x805ff89 in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1595
\#4  0x805fb6b in perl\_run \(my\_perl=0x816ae68\) at perl\.c&#8203;:1516
\#5  0x805c912 in main \(argc=4\, argv=0xbffff804\, env=0xbffff818\)
&nbsp;    at perlmain\.c&#8203;:76
\(gdb\) x sv
0x2&#8203;:    Cannot access memory at address 0x2

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.7.2:

Configured by john at Sun Dec  9 23:05:55 CET 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 2 patch 13539) configuration:
  Platform:
    osname=linux, osvers=2.2.19, archname=i686-linux
    uname='linux pc.xs4all.nl 2.2.19 #1 sun jun 24 15:06:07 cest 2001 i686 unknown '
    config_args='-Doptimize=-g -Dusedevel -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=define
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL13539


@INC for perl v5.7.2:
    lib
    /usr/local/lib/perl5/5.7.2/i686-linux
    /usr/local/lib/perl5/5.7.2
    /usr/local/lib/perl5/site_perl/5.7.2/i686-linux
    /usr/local/lib/perl5/site_perl/5.7.2
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.7.2:
    HOME=/usr/home/john
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/home/john/bin:/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/X11R6/bin:/usr/games:/usr/local/sbin:/usr/local/pgsql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @vanstyn

Jan\-Pieter Cornet \<john@&#8203;pc\.xs4all\.nl> wrote&#8203;:
:In some circumstances\, perl dumps core after executing code embedded
:in a regular expression\, with \(?\{ code\.\.\. \}\)\.

This is sorta known\. I just dug some more\, and it turns out that
this occurs because we switch curpads when performing the eval block
\(in regmatch\(\)\, case EVAL\)\, but because the eval isn't protected by
an ENTER/LEAVE we've left stuff on the stack referring to the wrong
curpad\. Using a shorter example\, I did this&#8203;:

&nbsp;  \(gdb\) set args \-we 'our$t="";""=~/\(?\{print "\<$t>\\n";my@&#8203;lt=\(\)\}\)/'
&nbsp;  \(gdb\) break Perl\_leave\_scope
&nbsp;  Breakpoint 1 at 0x80db059&#8203;: file scope\.c\, line 634\.
&nbsp;  \(gdb\) cond 1 base == 3
&nbsp;  \(gdb\) run
&nbsp;  \<>

&nbsp;  Breakpoint 1\, Perl\_leave\_scope \(base=3\) at scope\.c&#8203;:634
&nbsp;  634         if \(base \< \-1\)
&nbsp;  \(gdb\) p \(\(void\*\*\)PL\_savestack\)\[0\]@&#8203;PL\_savestack\_ix
&nbsp;  $1 = \{0xffffffff\, 0x8133b38\, 0x4\, 0x0\, 0x8138cdc\, 0x6\, 0xffffffff\, 0x8133b38\, 
&nbsp;    0x4\, 0x8137380\, 0x813cd48\, 0x8\, 0x8102980\, 0x0\, 0x1e\, 0x5\, 0x12\, 0x81373bc\, 
&nbsp;    0x81373c8\, 0x1\}
&nbsp;  \(gdb\) 

\.\.\. in which the \<\< 0x5\, 0x12 >> is a SAVEt\_CLEARSV of curpad\[5\]\. By the
time we finally LEAVE\_SCOPE at the end of pp\_match\(\)\, the curpad is
restored to something which doesn't point to an SV at \[5\]\.

Ilya strongly objected to an ENTER/LEAVE in re\_eval processing on speed
grounds\, but there is another problem&#8203;: we need localised variables to
retain their value for later in the regexp\. This is demonstrated by the
regexp test \#469&#8203;:
&nbsp;  \(?\{$a=2\}\)a\*aa\(?\{local$a=$a\+1\}\)k\*c\(?\{$b=$a\}\) yaaxxaaaacd y   $b  3
\.\.\. which expects $a == 3 in the third eval block\, and indeed adding an
ENTER/LEAVE pair causes re\_tests \#469 and \#470 to fail\, and no other\.

Now\, the curpad shenanigans here have always been a mess\, and I don't
know of an easy solution\. It may be that we can fix the curpad once
at the start of any regexp that includes an eval\, and do a single
ENTER/LEAVE for the entire regexp before restoring; I'm not quite sure
what's involved with that \- Sarathy might have more answers\.

That does suggest you have a better chance if you avoid using pad
variables within the eval\, and indeed if I move the my\(\) declaration
of $t and @&#8203;lt from your original test case out to the enclosing sub
it no longer cores\. I don't know how much of a guarantee that is \.\.\.

Hugo

**Comment by p5pRT** [comment]
### From @cpansprout

This may be another manifestation of bug \#8027\.

I have observed this in 5\.8\.8/9\, 5\.10\.0 and 5\.11\.0@&#8203;34471\. The  
following program prints ‘undef’\, when it should print ‘self’&#8203;:

\#\!/usr/local/bin/perl5\.10\.0

use warnings; use strict;

\#my $foo;
my $self = "self";

"i" =~ /\(?\{  my @&#8203;m  \}\)/;

warn "Should be self&#8203;: "\, $self // "undef";


It seems that the ‘my @&#8203;m’ in the regular expression is overwriting  
$self\. If I uncomment ‘my $foo’\, then $self is left untouched\.

I have two workarounds&#8203;: \(?\{\{ \.\.\. \}\}\) and \(?\{do\{\{ \.\.\. \}\}\}\)\. I’m not  
sure why those work\, as \(?\{do\{\.\.\.\}\}\) and \{"i" =~ /\(?\{\.\.\.\}\)/\} do not\.



---
Flags&#8203;:
&nbsp;     category=core
&nbsp;     severity=medium


---
This perlbug was built using Perl v5\.8\.8 \- Tue Jan 13 10&#8203;:01&#8203;:03 PST 2009
It is being executed now by  Perl v5\.10\.0 \- Sun Dec 23 12&#8203;:08&#8203;:07 PST  
2007\.

Site configuration information for perl v5\.10\.0&#8203;:

Configured by sprout at Sun Dec 23 12&#8203;:08&#8203;:07 PST 2007\.

Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
&nbsp;   Platform&#8203;:
&nbsp;     osname=darwin\, osvers=9\.0\.0\, archname=darwin\-2level
&nbsp;     uname='darwin pint\.local 9\.0\.0 darwin kernel version 9\.0\.0&#8203;: tue  
oct 9 21&#8203;:35&#8203;:55 pdt 2007; root&#8203;:xnu\-1228~1release\_i386 i386 '
&nbsp;     config\_args='\-de'
&nbsp;     hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;     useithreads=undef\, usemultiplicity=undef
&nbsp;     useperlio=define\, d\_sfio=undef\, uselargefiles=define\,  
usesocks=undef
&nbsp;     use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;     usemymalloc=n\, bincompat5005=undef
&nbsp;   Compiler&#8203;:
&nbsp;     cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-no\-cpp\-precomp \-fno\- 
strict\-aliasing \-pipe \-I/usr/local/include'\,
&nbsp;     optimize='\-O3'\,
&nbsp;     cppflags='\-no\-cpp\-precomp \-fno\-common \-DPERL\_DARWIN \-no\-cpp\- 
precomp \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;     ccversion=''\, gccversion='4\.0\.1 \(Apple Inc\. build 5465\)'\,  
gccosandvers=''
&nbsp;     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\,  
Off\_t='off\_t'\, lseeksize=8
&nbsp;     alignbytes=8\, prototype=define
&nbsp;   Linker and Libraries&#8203;:
&nbsp;     ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-L/usr/ 
local/lib'
&nbsp;     libpth=/usr/local/lib /usr/lib
&nbsp;     libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;     perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;     libc=/usr/lib/libc\.dylib\, so=dylib\, useshrplib=false\,  
libperl=libperl\.a
&nbsp;     gnulibc\_version=''
&nbsp;   Dynamic Linking&#8203;:
&nbsp;     dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;     cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/ 
usr/local/lib'

Locally applied patches&#8203;:
&nbsp;     fix for regcomp CVE\-2007\-5116 security vulnerability
&nbsp;     fix for double\-free CVE\-2008\-1927 security vulnerability



---
@&#8203;INC for perl v5\.10\.0&#8203;:
&nbsp;     /usr/local/lib/perl5/5\.10\.0/darwin\-2level
&nbsp;     /usr/local/lib/perl5/5\.10\.0
&nbsp;     /usr/local/lib/perl5/site\_perl/5\.10\.0/darwin\-2level
&nbsp;     /usr/local/lib/perl5/site\_perl/5\.10\.0
&nbsp;     /usr/local/lib/perl5/site\_perl/5\.8\.6
&nbsp;     /usr/local/lib/perl5/site\_perl
&nbsp;     \.



---
Environment for perl v5\.10\.0&#8203;:
&nbsp;     DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;     HOME=/Users/sprout
&nbsp;     LANG=en\_US\.UTF\-8
&nbsp;     LANGUAGE \(unset\)
&nbsp;     LD\_LIBRARY\_PATH \(unset\)
&nbsp;     LOGDIR \(unset\)
&nbsp;     PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/ 
usr/local/bin
&nbsp;     PERL\_BADLANG \(unset\)
&nbsp;     SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Sun Apr 26 12&#8203;:37&#8203;:42 2009\, sprout@&#8203;cpan\.org wrote&#8203;:
> This may be another manifestation of bug \#8027\.
> 
> I have observed this in 5\.8\.8/9\, 5\.10\.0 and 5\.11\.0@&#8203;34471\. The  
> following program prints ‘undef’\, when it should print ‘self’&#8203;:
> 
> \#\!/usr/local/bin/perl5\.10\.0
> 
> use warnings; use strict;
> 
> \#my $foo;
> my $self = "self";
> 
> "i" =~ /\(?\{  my @&#8203;m  \}\)/;
> 
> warn "Should be self&#8203;: "\, $self // "undef";
> 
> 
> It seems that the ‘my @&#8203;m’ in the regular expression is overwriting  
> $self\. If I uncomment ‘my $foo’\, then $self is left untouched\.
> 
> I have two workarounds&#8203;: \(?\{\{ \.\.\. \}\}\) and \(?\{do\{\{ \.\.\. \}\}\}\)\. I’m not  
> sure why those work\, as \(?\{do\{\.\.\.\}\}\) and \{"i" =~ /\(?\{\.\.\.\}\)/\} do not\.
> 
 

This is indeed the same bug as \#8027 except that \#8027 core dumps where 
your test case corrups memory \(basically the same thing\)\. \(So I'm 
merging the two tickets\)

Unfortunally there does not appear to be an easy fix :\(
\(As a side note&#8203;: this behaviour goes back to at least perl\-5\.00504\)


Best regards\,

Bram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rspier

manually unmerged with 38639

**Comment by p5pRT** [comment]
### From @cpansprout

This was fixed by commit 9e103e269\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
